paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/compact/changelog/ChangelogCompactTask.java

```diff
@@ -81,69 +91,71 @@ public Map<Integer, List<DataFileMeta>> compactChangelogFiles() {
         return compactChangelogFiles;
     }
 
-    public List<Committable> doCompact(FileStoreTable table) throws Exception {
+    public List<Committable> doCompact(FileStoreTable table, ExecutorService executor)
+            throws Exception {
         FileStorePathFactory pathFactory = table.store().pathFactory();
+        List<ReadTask> tasks = new ArrayList<>();
+        BiConsumer<Map<Integer, List<DataFileMeta>>, Boolean> addTasks =
+                (files, isCompactResult) -> {
+                    for (Map.Entry<Integer, List<DataFileMeta>> entry : files.entrySet()) {
+                        int bucket = entry.getKey();
+                        DataFilePathFactory dataFilePathFactory =
+                                pathFactory.createDataFilePathFactory(partition, bucket);
+                        for (DataFileMeta meta : entry.getValue()) {
+                            tasks.add(
+                                    new ReadTask(
+                                            table,
+                                            dataFilePathFactory.toPath(meta),
+                                            bucket,
+                                            isCompactResult,
+                                            meta));
+                        }
+                    }
+                };
+        addTasks.accept(newFileChangelogFiles, false);
+        addTasks.accept(compactChangelogFiles, true);
+
+        Iterator<ReadTask> it =
+                ThreadPoolUtils.randomlyExecuteSequentialReturn(
+                        executor,
+                        t -> {
+                            // Total lengths of all bytes will not exceed `targetFileSize * 2`,
```

Comment:
I feel it is better to use workers and queue and consumer. Even max `targetFileSize * 2`, if target file size is 1GB, this is too still large.

Workers and queue is safer.